### ISSUE TITLE
What does Docker for Mac use for DNS resolution?

### ISSUE THREAD

**Issue by ntwrkguru** [docker/for-mac#1348]
### Expected behavior
Docker for Mac uses resolv.conf for DNS resolution

### Actual behavior
Docker for Mac uses something else for DNS resolution

### Information
	
What mechanism does Docker for Mac use to resolve DNS? I am unable to log in to any private registries due to DNS issues, but the resolver shouldn't be used as I don't see it in resolv.conf or scutil. So that I can track down where the issue is, I need to understand how Docker for Mac resolves DNS.

```
Error response from daemon: Get https://reg.example.net:5000/v1/users/: dial tcp: lookup reg.example.net on 192.168.65.1:53: no such host
```

```
$ ifconfig | grep 192.168.65
$ cat /etc/resolv.conf | grep 192.168.65
$ cat /etc/hosts | grep 192.168.65
$ scutil --dns | grep 192.168.65
```

### Steps to reproduce the behavior

  1. Install Docker for Mac
  2. Attempt to connect to a private registry
  3. Open a case with Docker when it doesn't work and get told you have to use community support
  4. Hope for community support even though you pay for Docker's Enterprise products
  5. Don't renew your subscription to Docker Datacenter/CS Engine
  6. Start searching for a better alternative to Docker for Mac

**Comment by ntwrkguru** [comment]
This is confirmed to work as expected in docker-machine:

```
% docker login reg.example.net:5000
Username: user
Password: 
Login Succeeded
```

**Comment by LongLiveCHIEF** [comment]
likely related to: https://github.com/docker/docker/issues/22907 (which was never fixed, but closed anyways).

I'm experiencing this issue as well, _exact_ same info/specs.

**Comment by ntwrkguru** [comment]
@LongLiveCHIEF agreed. And setting a proxy is **not** the answer. If Docker for Mac runs as a "native application", it should use the the same name resolution of any other Mac application.

**Comment by damau** [comment]
We also have this problem.  Would really like to see a fix.

**Comment by ntwrkguru** [comment]
Appears to be fixed in 17.03.0-ce-mac1 (15583), but I'd like to know which RLI resolved it. Looking at the [release notes](https://docs.docker.com/docker-for-mac/release-notes/#stable-release-notes) there are numerous DNS bug fixes, but not sure which one fixes this.

**Comment by djs55** [comment]
@ntwrkguru thanks for letting us know it's fixed in 17.03.0-ce-mac1.

Docker for Mac and Docker for Windows use a built-in DNS proxy which listens on the internal IP `192.168.65.1`. Linux resolvers don't usually support a configuration as sophisticated as that on the Mac or Windows -- for example there is often a limit of 3 upstream servers and they don't usually support split-DNS used by VPN software. The DNS proxy in D4M/D4W tries to bridge this gap -- the container sends all request to `192.168.65.1` and the proxy sends it to

- servers found in `/etc/resolver/...`
- servers found in `scutil`

Note that `/etc/resolv.conf` on macOS now includes the notice
```
# Mac OS X Notice
#
# This file is not used by the host name and address resolution
# or the DNS query routing mechanisms used by most processes on
# this Mac OS X system.
#
# This file is automatically generated.
```
-- I believe this file is automatically generated from the contents stored in the SC database, so D4M reads the values from the SC database directly, since this also contains `SupplementalMatchDomains` (configuration for split DNS) and server order/priority information.

The actual upstream servers being used by D4M can be viewed by:

```
$ cd ~/Library/Containers/com.docker.docker/Data/database/
$ git reset --hard
HEAD is now at c607869 last-start-time changed at 1489071742
$ cat com.docker.driver.amd64-linux/slirp/dns 
# { Addresses: 10.14.32.10; Order: 200000; Zones:  }
nameserver 10.14.32.10
order 200000
timeout 2000
```
(Note there are sometimes duplicates in this file but this should be harmless)

Regarding which change log line item might have fixed the problem: unfortunately without the original diagnostics I can't tell for certain. There were numerous bug fixes but unfortunately most (if not all) DNS problems manifest themselves as an error like `no such host`.

Thanks again for your report and for confirming the bug has gone!

**Comment by ntwrkguru** [comment]
Thanks @djs55.

**Comment by parkan** [comment]
FWIW this seems to have been fixed in `17.04.0-ce-mac7` for me, not in `17.03.*`

**Comment by bjbrewster** [comment]
I just had a similar issue to this after upgrading to latest Docker on 7th July 2017.  We are using a private registry on our own network.  `/etc/resolv.conf` and `scutil --dns` shows our two DNS nameservers and correct search domain but this seemed to be ignored by Docker.  I had manually added the DNS settings ("dns" and "dns-search") in the Docker Preferences -> Advanced tab a while ago so it seemed unrelated to Docker using the system DNS settings.

I updated to the latest Docker `17.06.0-ce-mac19 (18663)` today (17th July 2017) and it now works fine as before my previous upgrade.  I'm not sure which version I was on before.
